File: AulaTPII/F217/GoF/Estruturais/Proxy/Proxy01.js
// Objeto Real
class ServicoReal{
    executar(){
        console.log("Executando servidor real")
    }
}

// Proxy
class ProxyServico{
    constructor(){
        this.servicoReal = new ServicoReal();
    }

    executar() {
        console.log("Antes da execução do serviço");
        this.servicoReal.executar();
        console.log("Depois da execução do serviço");
    }
}
// USO DO PROXY - CLIENTE ##########################
const proxy = new ProxyServico();
proxy.executar();

File: AulaTPII/F217/GoF/Estruturais/Proxy/Proxy01.java
// Objeto Real
class ServicoReal{
    public void executar(){
        System.out.println("Executando servidor real");
    }
}

// Proxy
class ProxyServico{
    private ServicoReal servicoReal;

    public ProxyServico(){
        this.servicoReal = new ServicoReal();
    }

    public void executar(){
        System.out.println("Antes da execução do serviço");
        servicoReal.executar();
        System.out.println("Depois da Execução do Serviço");
    }
}

// CLIENTE - USO DO PROXY
public class Proxy01{
    public static void main(String[] args){
        ProxyServico proxy = new ProxyServico();
        proxy.executar();
    }
}